R/assess_dependencies.R

```diff
@@ -0,0 +1,121 @@
+assess_dependencies.pkg_cran_remote <- function(x, ...){
+  pkg_metric_eval(class = "pkg_metric_dependencies", {
+      get_package_dependencies(x$name, repo = getOption("repos")[["CRAN"]])
+  })
+}
+
+#' @importFrom BiocManager repositories
+#' @export
+assess_dependencies.pkg_bioc_remote <- function(x, ...){
+  pkg_metric_eval(class = "pkg_metric_dependencies", {
+    get_package_dependencies(x$name, BiocManager::repositories()[1])
+  })
+}
+
+#' Score a package for dependencies
+#'
+#' Returns the total number dependencies
+#'
+#' @eval roxygen_score_family("dependencies")
+#' @return A \code{numeric}
+#'
+#' @export
+metric_score.pkg_metric_dependencies <- function(x, ...) {
+  NROW(x)
+}
+
+attributes(metric_score.pkg_metric_dependencies)$label <-
+  "The number of package dependencies"
+
+#Helper functions to get extract dependencies
+
+#' Gets available packages from necessary repository and filters for
+#' package of interest
+#'
+#' @param name package name
+#' @param repo package repository (e.g. CRAN or Bioconductor)
+#'
+#' @return Returns a data frame with two columns 1) Package names, 2) type of dependency (LinkingTo, Imports, Depends)
+#'
+get_package_dependencies <- function(name, repo){
+  ap <- available.packages(repos = repo)
+  deps <- ap[rownames(ap)==name, c("LinkingTo","Imports","Depends")]
+  deps <- deps[!is.na(deps)]
+  deps <- sapply(strsplit(deps, ","), trimws)
+  deps <- data.frame(package=unlist(deps),
+                     type=rep(names(deps), sapply(deps, length)),
+                     stringsAsFactors = FALSE,
+                     row.names = NULL)
+  deps <- remove_base_packages(deps)
+  return(deps)
+}
+
+#' Parse DCF of description file
+#'
+#' @param path pkg_ref path
+#'
+parse_dcf_dependencies <- function(path){
+  dcf <- read.dcf(file.path(path, "DESCRIPTION"), all=TRUE)
+  dcf <- dcf[colnames(dcf) %in% c("LinkingTo","Imports", "Depends")]
+  dcf <- sapply(dcf, strsplit, strsplit, split=",")
+  dcf <- lapply(dcf, trimws)
+  deps <- data.frame(package=unlist(dcf),
+                     type=rep(names(dcf), sapply(dcf, length)),
+                     stringsAsFactors = FALSE,
+                     row.names = NULL)
+  deps <- remove_base_packages(deps)
+  return(deps)
+}
+
+#' Helper function to remove base and recommended packages
+#'
+#' @param df Data frame of dependencies of a package.
+#'
+remove_base_packages <- function(df){
+  deps <- df[!grepl("^R\\s\\(.+\\)", df$package) | df$package %in% c('base','compiler','datasets',
+                                                                           'graphics','grDevices','grid',
+                                                                           'methods','parallel','splines',
+                                                                           'stats','stats4','tcltk','tools',
+                                                                           'utils'), ] ##Remove base R dependencies and in the future recommended pacakges
```

Comment:
Instead of hardcoding, we can use `installed.packages()[,"Priority"]` to get a list of base/recommended packages:

```r
inst <- memoise_installed_packages()
inst_prio <- inst[,"Priority"]
inst_is_base_rec <- !is.na(inst_prio) & inst_prio %in% c("base", "recommended")
base_rec_pkgs <- inst[inst_is_base_rec, "Package"]
```